File: oficina-sistema/backend/app/services/pdf.py
"""Geração do PDF de orçamento de uma Ordem de Serviço, para envio ao
cliente. Implementação enxuta com fpdf2 (puro Python, sem dependência de
sistema) — formatação rica (logo, cabeçalho da oficina, etc.) fica para uma
iteração futura quando houver identidade visual definida.
"""

from decimal import Decimal

from fpdf import FPDF

from app.models.cliente import Cliente
from app.models.ordem_servico import OrdemServico
from app.models.veiculo import Veiculo


def _moeda(valor: Decimal) -> str:
    return f"R$ {valor:,.2f}".replace(",", "_").replace(".", ",").replace("_", ".")


def gerar_pdf_orcamento(
    os_: OrdemServico,
    cliente: Cliente,
    veiculo: Veiculo,
    itens_peca: list[dict],
    valor_total: Decimal,
) -> bytes:
    """`itens_peca` é uma lista de dicts já com a descrição da peça
    resolvida (join feito no router): {descricao, quantidade,
    preco_unitario_venda, subtotal}."""

    pdf = FPDF()
    pdf.add_page()
    pdf.set_font("Helvetica", "B", 16)
    pdf.cell(0, 10, f"Orcamento - OS #{os_.numero}", ln=True)

    pdf.set_font("Helvetica", "", 10)
    pdf.cell(0, 6, f"Status: {os_.status}", ln=True)
    pdf.cell(0, 6, f"Data de abertura: {os_.data_abertura.strftime('%d/%m/%Y')}", ln=True)
    if os_.prazo_estimado:
        pdf.cell(0, 6, f"Prazo estimado: {os_.prazo_estimado.strftime('%d/%m/%Y')}", ln=True)
    if os_.forma_pagamento:
        pdf.cell(0, 6, f"Forma de pagamento: {os_.forma_pagamento}", ln=True)
    pdf.ln(4)

    pdf.set_font("Helvetica", "B", 12)
    pdf.cell(0, 8, "Cliente", ln=True)
    pdf.set_font("Helvetica", "", 10)
    pdf.cell(0, 6, f"Nome: {cliente.nome}", ln=True)
    if cliente.telefone:
        pdf.cell(0, 6, f"Telefone: {cliente.telefone}", ln=True)
    if cliente.cpf_cnpj:
        pdf.cell(0, 6, f"CPF/CNPJ: {cliente.cpf_cnpj}", ln=True)
    pdf.ln(2)

    pdf.set_font("Helvetica", "B", 12)
    pdf.cell(0, 8, "Veiculo", ln=True)
    pdf.set_font("Helvetica", "", 10)
    veic_linha = " ".join(filter(None, [veiculo.marca, veiculo.modelo, str(veiculo.ano or "")]))
    pdf.cell(0, 6, f"{veic_linha} - Placa {veiculo.placa}", ln=True)
    pdf.ln(4)

    if itens_peca:
        pdf.set_font("Helvetica", "B", 12)
        pdf.cell(0, 8, "Pecas", ln=True)
        pdf.set_font("Helvetica", "B", 9)
        pdf.cell(90, 6, "Descricao", border="B")
        pdf.cell(25, 6, "Qtd.", border="B", align="R")
        pdf.cell(35, 6, "Preco unit.", border="B", align="R")
        pdf.cell(35, 6, "Subtotal", border="B", align="R", ln=True)
        pdf.set_font("Helvetica", "", 9)
        for item in itens_peca:
            pdf.cell(90, 6, str(item["descricao"])[:48])
            pdf.cell(25, 6, str(item["quantidade"]), align="R")
            pdf.cell(35, 6, _moeda(item["preco_unitario_venda"]), align="R")
            pdf.cell(35, 6, _moeda(item["subtotal"]), align="R", ln=True)
        pdf.ln(4)

    if os_.itens_servico:
        pdf.set_font("Helvetica", "B", 12)
        pdf.cell(0, 8, "Servicos", ln=True)
        pdf.set_font("Helvetica", "B", 9)
        pdf.cell(150, 6, "Descricao", border="B")
        pdf.cell(35, 6, "Valor", border="B", align="R", ln=True)
        pdf.set_font("Helvetica", "", 9)
        for item in os_.itens_servico:
            pdf.cell(150, 6, str(item.descricao)[:80])
            pdf.cell(35, 6, _moeda(item.valor), align="R", ln=True)
        pdf.ln(4)

    pdf.set_font("Helvetica", "B", 13)
    pdf.cell(0, 10, f"Total: {_moeda(valor_total)}", ln=True)

    return bytes(pdf.output())

File: oficina-sistema/backend-next/lib/pdf.ts
import { PDFDocument, StandardFonts, rgb, type PDFFont, type PDFPage } from "pdf-lib";

/** Geração do PDF de orçamento de uma OS, pra envio ao cliente. Réplica em
 * JS puro (pdf-lib, sem dependência de sistema — igual o espírito do
 * fpdf2 usado no FastAPI antigo) do layout de `app/services/pdf.py`:
 * cabeçalho, cliente, veículo, tabela de peças, tabela de serviços, total. */

function moeda(valor: number): string {
  return `R$ ${valor.toFixed(2).replace(".", ",")}`;
}

function dataPtBr(iso: string): string {
  const d = new Date(iso);
  return d.toLocaleDateString("pt-BR", { timeZone: "UTC" });
}

interface ItemPecaPdf {
  descricao: string;
  quantidade: number;
  preco_unitario_venda: number;
  subtotal: number;
}

interface ItemServicoPdf {
  descricao: string;
  valor: number;
}

export async function gerarPdfOrcamento(params: {
  numero: number;
  status: string;
  dataAbertura: string;
  prazoEstimado: string | null;
  formaPagamento: string | null;
  clienteNome: string;
  clienteTelefone: string | null;
  clienteCpfCnpj: string | null;
  veiculoLinha: string;
  itensPeca: ItemPecaPdf[];
  itensServico: ItemServicoPdf[];
  valorTotal: number;
}): Promise<Uint8Array> {
  const pdfDoc = await PDFDocument.create();
  let page: PDFPage = pdfDoc.addPage([595.28, 841.89]); // A4
  const margem = 40;
  let y = 841.89 - margem;
  const font: PDFFont = await pdfDoc.embedFont(StandardFonts.Helvetica);
  const fontBold: PDFFont = await pdfDoc.embedFont(StandardFonts.HelveticaBold);
  const preto = rgb(0.1, 0.1, 0.1);
  const cinza = rgb(0.4, 0.4, 0.4);

  function linha(texto: string, opts: { negrito?: boolean; tamanho?: number; cor?: typeof preto } = {}) {
    if (y < margem + 20) {
      page = pdfDoc.addPage([595.28, 841.89]);
      y = 841.89 - margem;
    }
    page.drawText(texto, {
      x: margem,
      y,
      size: opts.tamanho ?? 10,
      font: opts.negrito ? fontBold : font,
      color: opts.cor ?? preto,
    });
    y -= (opts.tamanho ?? 10) + 6;
  }

  linha(`Orçamento — OS #${params.numero}`, { negrito: true, tamanho: 18 });
  y -= 4;
  linha(`Status: ${params.status}`, { cor: cinza });
  linha(`Data de abertura: ${dataPtBr(params.dataAbertura)}`, { cor: cinza });
  if (params.prazoEstimado) linha(`Prazo estimado: ${dataPtBr(params.prazoEstimado)}`, { cor: cinza });
  if (params.formaPagamento) linha(`Forma de pagamento: ${params.formaPagamento}`, { cor: cinza });
  y -= 6;

  linha("Cliente", { negrito: true, tamanho: 12 });
  linha(`Nome: ${params.clienteNome}`);
  if (params.clienteTelefone) linha(`Telefone: ${params.clienteTelefone}`);
  if (params.clienteCpfCnpj) linha(`CPF/CNPJ: ${params.clienteCpfCnpj}`);
  y -= 4;

  linha("Veículo", { negrito: true, tamanho: 12 });
  linha(params.veiculoLinha);
  y -= 8;

  if (params.itensPeca.length > 0) {
    linha("Peças", { negrito: true, tamanho: 12 });
    linha(
      `${"Descrição".padEnd(40)}${"Qtd.".padStart(8)}${"Preço unit.".padStart(14)}${"Subtotal".padStart(14)}`,
      { negrito: true, tamanho: 9 }
    );
    for (const item of params.itensPeca) {
      linha(
        `${item.descricao.slice(0, 40).padEnd(40)}${String(item.quantidade).padStart(8)}` +
          `${moeda(item.preco_unitario_venda).padStart(14)}${moeda(item.subtotal).padStart(14)}`,
        { tamanho: 9 }
      );
    }
    y -= 6;
  }

  if (params.itensServico.length > 0) {
    linha("Serviços", { negrito: true, tamanho: 12 });
    linha(`${"Descrição".padEnd(60)}${"Valor".padStart(14)}`, { negrito: true, tamanho: 9 });
    for (const item of params.itensServico) {
      linha(`${item.descricao.slice(0, 60).padEnd(60)}${moeda(item.valor).padStart(14)}`, { tamanho: 9 });
    }
    y -= 6;
  }

  linha(`Total: ${moeda(params.valorTotal)}`, { negrito: true, tamanho: 14 });

  return pdfDoc.save();
}
